optimize/client/src/modules/components/Tabs/Tabs.tsx

```diff
@@ -6,74 +6,117 @@
  * except in compliance with the Camunda License 1.0.
  */
 
-import {useState, useEffect, ComponentPropsWithoutRef, ReactNode} from 'react';
+import {
+  useState,
+  ComponentPropsWithoutRef,
+  ReactNode,
+  createContext,
+  useContext,
+  cloneElement,
+  ReactElement,
+} from 'react';
 import {Tabs as CarbonTabs, TabList, Tab, TabPanels, TabPanel, TabsSkeleton} from '@carbon/react';
 
 import {ignoreFragments} from 'services';
 
 import './Tabs.scss';
 
-interface TabsProps<T extends string | number>
-  extends Omit<ComponentPropsWithoutRef<'div'>, 'onChange'> {
+type TabValue = number | string;
+
+const TabsContext = createContext<{
+  alreadyOpened: TabValue[];
+}>({
+  alreadyOpened: [],
+});
+
+const useTabsContext = () => {
+  const context = useContext(TabsContext);
+  if (!context) {
+    throw new Error('useTabsContext must be used within a Tabs component');
+  }
+  return context;
+};
+
+interface TabsProps<T extends TabValue> extends Omit<ComponentPropsWithoutRef<'div'>, 'onChange'> {
   value?: T;
   onChange?: (value: T) => void;
   showButtons?: boolean;
   isLoading?: boolean;
 }
 
-interface TabProps<T extends string | number>
-  extends Omit<ComponentPropsWithoutRef<'div'>, 'title'> {
+interface TabProps<T extends TabValue> extends Omit<ComponentPropsWithoutRef<'div'>, 'title'> {
   value?: T;
   title?: ReactNode;
   disabled?: boolean;
 }
 
-export default function Tabs<T extends string | number>({
-  value = 0 as T,
+export default function Tabs<T extends TabValue>({
+  value,
   onChange,
   children,
   showButtons = true,
   isLoading,
 }: TabsProps<T>) {
-  const tabs = ignoreFragments(children);
-  const values = tabs.map<T>(({props: {value}}, idx) => (value || idx) as T);
-  const [selected, setSelected] = useState<T>(value);
+  const tabs = ignoreFragments<TabProps<T>>(children);
+  const values = tabs.map<T>(({props: {value}}, idx) => value ?? (idx as T));
+  const initialValue = value ?? values[0] ?? 0;
+  const [selectedIndex, setSelectedIndex] = useState<number>(getIndex(values, initialValue));
+  const [alreadyOpened, setAlreadyOpened] = useState<TabValue[]>([initialValue]);
+
+  function handleTabChange({selectedIndex}: {selectedIndex: number}) {
+    const newValue = values[selectedIndex];
+
+    if (newValue === undefined) {
+      return;
+    }
 
-  useEffect(() => {
-    setSelected(value);
-  }, [value]);
+    onChange?.(newValue);
+    setSelectedIndex(selectedIndex);
+
+    if (!alreadyOpened.includes(newValue)) {
+      setAlreadyOpened((prev) => [...prev, newValue]);
```

Comment:
Previously this was cached implicitly as well, cause all the tabs were loaded once and then not re-rendered as far as I can see. I checked on stage, following the steps described and this bug is present there. If we are okay with re-rendering (and re-requesting APIs) each time we open a tab, then I can remove this caching mechanism, but I think the issue we want to solve is to reduce the amount of API calls.

Maybe we should think of some local context there which we update so the reports list is always up to date?

Reply:
Maybe we should just re-fetch reports on the alert modal open?

Reply:
> Previously this was cached implicitly as well, cause all the tabs were loaded once and then not re-rendered as far as I can see.
I think we broke it when we migrated to carbon. Our previous tabs component used to only load the visible tabs.

> I think the issue we want to solve is to reduce the amount of API calls.
The issue we are trying to solve is to prevent issuing API calls that we might never need. Re-issuing request when a user revisit a page is completely fine in my opinion.
We currently follow a stateless API approach everywhere in the codebase since caching is not easy to do without a library.

Reply:
so what is the resolution now? Just adding a prop that allows to load all tabs at once?